subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -102,61 +103,83 @@ def check_payments(self, payments: Iterable[SubscriptionPayment]):
         with suppress(SubscriptionPayment.DoesNotExist):  # noqa (DoesNotExist seems not to be an exception for PyCharm)
             payment = SubscriptionPayment.objects.get(provider_codename=self.codename,
-                                                      provider_transaction_id=single_in_app.transaction_id)
-            # We've already handled this. And all the messages coming to us from iOS should be AFTER
-            # the client money were removed.
-            return Response(
-                SubscriptionPaymentSerializer(payment).data,
-                status=HTTP_200_OK,
-            )
+                                                      provider_transaction_id=receipt_info.transaction_id)
+
+            # User was refunded.
+            if receipt_info.cancellation_date is not None:
+                payment.subscription_end = receipt_info.cancellation_date
+                payment.save()
+
+            return payment
 
         # Find the right plan to create subscription.
         try:
             search_kwargs = {
-                f'metadata__{self.codename}': single_in_app.product_id
+                f'metadata__{self.codename}': receipt_info.product_id
             }
             plan = Plan.objects.get(**search_kwargs)
         except Plan.DoesNotExist:
-            logger.warning('Plan for apple in-app purchase "%s" not found.', single_in_app.product_id)
-            return Response(status=HTTP_404_NOT_FOUND)
+            # This means that something wasn't connected as needed.
+            logger.exception('Plan for apple in-app purchase "%s" not found.', receipt_info.product_id)
+            return None
 
         # Create subscription payment. Subscription is created automatically.
         subscription_payment = SubscriptionPayment.objects.create(
             provider_codename=self.codename,
-            provider_transaction_id=single_in_app.transaction_id,
+            provider_transaction_id=receipt_info.transaction_id,
+            # NOTE(kkalinowski): from my understanding, we can only receive receipt after the person has paid
+            # but the transaction (from the perspective of the app) might have not yet finished. Money were spent.
             status=SubscriptionPayment.Status.COMPLETED,
             # In-app purchase doesn't report the money.
             # We mark it as None to indicate we don't know how much did it cost.
             amount=None,
-            user=request.user,
+            user=user,
             plan=plan,
-            subscription_start=single_in_app.purchase_date,
-            subscription_end=single_in_app.expires_date,
+            subscription_start=receipt_info.purchase_date,
+            # If the cancellation date is set, it means that the user was refunded and for whatever reason
+            # we weren't notified about this purchase.
+            subscription_end=receipt_info.cancellation_date or receipt_info.expires_date,
```

Comment:
I don't fully understand, but in general, if for whatever reason we weren't notified about something that we should be notified about, then maybe emit a warning?

Reply:
Raising an error if we have a cancellation – this will require a bit of rework.